From 9afd783fd50685cbde82dce89d02edf781f26470 Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Date: Wed, 8 Feb 2023 15:03:02 +0530
Subject: [PATCH] ath12k: fix qmi handler terminator entry

Since there is no terminator entry for ath12k_qmi_msg_handlers facing
below KASAN warning,

[   20.006340] ==================================================================
[   20.006385] BUG: KASAN: global-out-of-bounds in qmi_invoke_handler+0xa4/0x148
[   20.012453] Read of size 8 at addr ffffffd00a6428d8 by task kworker/u8:2/1273
[   20.019652]
[   20.026773] CPU: 0 PID: 1273 Comm: kworker/u8:2 Not tainted 5.4.213 #0
[   20.028334] Hardware name: Qualcomm Technologies, Inc. IPQ9574/AP-AL02-C4 (DT)
[   20.034681] Workqueue: qmi_msg_handler qmi_data_ready_work
[   20.041873] Call trace:
[   20.047347]  dump_backtrace+0x0/0x20c
[   20.049690]  show_stack+0x14/0x1c
[   20.053509]  dump_stack+0xe0/0x138
[   20.056808]  print_address_description.isra.5+0x30/0x330
[   20.060106]  __kasan_report+0x16c/0x1bc
[   20.065573]  kasan_report+0xc/0x14
[   20.069132]  __asan_load8+0xa8/0xb0
[   20.072605]  qmi_invoke_handler+0xa4/0x148
[   20.075990]  qmi_handle_message+0x18c/0x1bc
[   20.080157]  qmi_data_ready_work+0x4ec/0x528
[   20.084242]  process_one_work+0x2c0/0x440
[   20.088753]  worker_thread+0x324/0x4b8
[   20.092657]  kthread+0x210/0x228
[   20.096302]  ret_from_fork+0x10/0x18
[   20.099682]
[   20.103246] The address belongs to the variable:
[   20.104827]  ath12k_mac_mon_status_filter_default+0x4bd8/0xfffffffffffe2300 [ath12k]
[   20.109668]
[   20.117563] Memory state around the address:
[   20.119049]  ffffffd00a642780: 00 00 00 00 00 00 02 fa fa fa fa fa 00 00 00 00
[   20.123650]  ffffffd00a642800: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[   20.130856] >ffffffd00a642880: 00 00 00 00 00 00 00 00 fa fa fa fa 00 00 00 00
[   20.138056]                                                     ^
[   20.145264]  ffffffd00a642900: 00 00 fa fa fa fa fa fa 00 00 00 00 01 fa fa fa
[   20.151428]  ffffffd00a642980: fa fa fa fa 00 00 00 fa fa fa fa fa 00 00 00 06
[   20.158542] ==================================================================

Fixed by adding NULL terminator entry to help qmi_invoke_handler api to
traverse upto the terminator entry without accessing out-of-boundary
index

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/qmi.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/qmi.c b/drivers/net/wireless/ath/ath12k/qmi.c
index 78887dcfdb07..5abc899b6d96 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -5001,6 +5001,8 @@ static const struct qmi_msg_handler ath12k_qmi_msg_handlers[] = {
 				sizeof(struct qmi_wlanfw_qdss_trace_save_ind_msg_v01),
 		.fn = ath12k_wlfw_qdss_trace_save_ind_cb,
 	},
+	/* (Additions here) */
+	{ /* terminator entry */ }
 };
 
 static int ath12k_qmi_ops_new_server(struct qmi_handle *qmi_hdl,
-- 
2.38.0

